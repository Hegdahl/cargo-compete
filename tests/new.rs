pub mod common;

use ignore::overrides::Override;
use insta::{assert_json_snapshot, assert_snapshot};
use liquid::object;
use std::io::BufRead;

#[test]
fn atcoder_agc047() -> anyhow::Result<()> {
    let (output, tree) = run(&b""[..], "agc047")?;
    assert_snapshot!("atcoder_agc047_output", output);
    assert_json_snapshot!("atcoder_agc047_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

#[cfg(feature = "__test_with_credentials")]
#[test]
fn atcoder_practice() -> anyhow::Result<()> {
    let (output, tree) = run(common::atcoder_credentials()?, "practice")?;
    assert_snapshot!("atcoder_practice_output", output);
    assert_json_snapshot!("atcoder_practice_file_tree", tree, { r#".**["Cargo.lock"]"# => ".." });
    Ok(())
}

fn run(
    input: impl BufRead + 'static,
    contest: &str,
) -> anyhow::Result<(String, serde_json::Value)> {
    common::run(
        |workspace_root| -> _ {
            std::fs::create_dir_all(workspace_root.join("cargo-compete-template").join("src"))?;

            std::fs::write(
                workspace_root.join("Cargo.toml"),
                r#"[workspace]
members = ["cargo-compete-template"]
"#,
            )?;

            std::fs::write(
                workspace_root.join("compete.toml"),
                liquid::ParserBuilder::with_stdlib()
                    .build()?
                    .parse(include_str!("../resources/compete.toml.liquid"))?
                    .render(&object!({
                        "template_platform": "atcoder",
                        "submit_via_binary": false,
                    }))?,
            )?;

            std::fs::write(
                workspace_root
                    .join("cargo-compete-template")
                    .join("Cargo.toml"),
                r#"[package]
name = "cargo-compete-template"
version = "0.1.0"
edition = "2018"
"#,
            )?;

            std::fs::write(
                workspace_root
                    .join("cargo-compete-template")
                    .join("src")
                    .join("main.rs"),
                r#"fn main() {
    todo!();
}
"#,
            )?;
            Ok(())
        },
        input,
        &["", "compete", "n", contest],
        |workspace_root, output| {
            output
                .replace(workspace_root.to_str().unwrap(), "{{ cwd }}")
                .replace('/', "{{ slash_or_backslash }}")
                .replace('\\', "{{ slash_or_backslash }}")
        },
        |_| Ok(Override::empty()),
    )
}
