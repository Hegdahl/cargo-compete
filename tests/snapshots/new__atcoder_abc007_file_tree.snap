---
source: tests/new.rs
expression: tree
---
{
  ".cargo": {
    "config.toml": "[cargo-new]\nname = \"\"\nemail = \"\"\n"
  },
  "abc007": {
    "Cargo.toml": "[package]\nname = \"abc007\"\nversion = \"0.1.0\"\nauthors = [\"\"]\nedition = \"2018\"\n\n[package.metadata.cargo-compete.bin]\na = { name = \"abc007-a\", problem = \"https://atcoder.jp/contests/abc007/tasks/abc007_1\" }\nb = { name = \"abc007-b\", problem = \"https://atcoder.jp/contests/abc007/tasks/abc007_2\" }\nc = { name = \"abc007-c\", problem = \"https://atcoder.jp/contests/abc007/tasks/abc007_3\" }\nd = { name = \"abc007-d\", problem = \"https://atcoder.jp/contests/abc007/tasks/abc007_4\" }\n\n[[bin]]\nname = \"abc007-a\"\npath = \"src/bin/a.rs\"\n\n[[bin]]\nname = \"abc007-b\"\npath = \"src/bin/b.rs\"\n\n[[bin]]\nname = \"abc007-c\"\npath = \"src/bin/c.rs\"\n\n[[bin]]\nname = \"abc007-d\"\npath = \"src/bin/d.rs\"\n[dependencies]\nproconio = \"=0.3.6\"\n",
    "src": {
      "bin": {
        "a.rs": "fn main() {\n    todo!();\n}\n",
        "b.rs": "fn main() {\n    todo!();\n}\n",
        "c.rs": "fn main() {\n    todo!();\n}\n",
        "d.rs": "fn main() {\n    todo!();\n}\n"
      }
    },
    "testcases": {
      "a.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      4\n    out: |\n      3\n  - name: sample2\n    in: |\n      100\n    out: |\n      99\n  - name: sample3\n    in: |\n      1\n    out: |\n      0\n\nextend: []\n",
      "b.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      xyz\n    out: |\n      xy\n  - name: sample2\n    in: |\n      c\n    out: |\n      b\n  - name: sample3\n    in: |\n      a\n    out: |\n      -1\n  - name: sample4\n    in: |\n      aaaaa\n    out: |\n      aaaa\n\nextend: []\n",
      "c.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      7 8\n      2 2\n      4 5\n      ########\n      #......#\n      #.######\n      #..#...#\n      #..##..#\n      ##.....#\n      ########\n    out: |\n      11\n  - name: sample2\n    in: |\n      5 8\n      2 2\n      2 4\n      ########\n      #.#....#\n      #.###..#\n      #......#\n      ########\n    out: |\n      10\n  - name: sample3\n    in: |\n      50 50\n      2 2\n      49 49\n      ##################################################\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      #................................................#\n      ##################################################\n    out: |\n      94\n\nextend: []\n",
      "d.yml": "---\ntype: Batch\ntimelimit: 2s\nmatch: Lines\n\ncases:\n  - name: sample1\n    in: |\n      1 9\n    out: |\n      2\n  - name: sample2\n    in: |\n      40 49\n    out: |\n      10\n  - name: sample3\n    in: |\n      1 1000\n    out: |\n      488\n  - name: sample4\n    in: |\n      1 1000000000000000000\n    out: |\n      981985601490518016\n\nextend: []\n"
    }
  },
  "compete.toml": "test-suite = \"{{ manifest_dir }}/testcases/{{ bin_alias | kebabcase }}.yml\"\n\n[new]\nplatform = \"atcoder\"\npath = \"./{{ package_name }}\"\n\n[new.template.dependencies]\nkind = \"inline\"\ncontent = '''\nproconio = \"=0.3.6\"\n'''\n\n[new.template.src]\nkind = \"inline\"\ncontent = '''\nfn main() {\n    todo!();\n}\n'''\n"
}
